Automattic/jetpack
# Update/auto redirect after install condition

Adds a method to detect plugin install context so that we can conditionally handle redirection after plugins are activated.

#### Changes proposed in this Pull Request:

* Adds a new function in `Plugins_Installer` to check the plugin installation context
* Adds a conditional check in the protect plugin that calls the new function before redirecting on plugin activation.

#### Other information:


#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



No

#### Testing instructions:






##### Reproduce the issue
* To see the original issue, you can test locally with the trunk branch checked out.
* First, try to enable the protect plugin from the CLI: `wp plugin activate protect`. If you are testing with the docker environment, you can run this using `jetpack docker wp plugin activate protect`.
* You will get a `Warning: Some code is trying to do a URL redirect.` message with an error backtrace. The protect plugin is still enabled, but the return is not clean.
* To test activating a plugin over the API, we can temporarily change the Creative Mail recommendation logic used to install the Creative Mail plugin in the recommendations flow. Update [this line](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/_inc/client/recommendations/feature-utils.js#L96) in `feature-utils.js` so that the plugin installed is `protect`.
* You will need to re-build the Jetpack plugin with `jetpack build plugins/jetpack`.
* You will also need to deactivate the protect plugin before the next step.
* Now, visit `/wp-admin/admin.php?page=jetpack#/recommendations/summary` and make sure your browser inspector console is open. Click the "Install" button next to the "Creative Mail" recommendation. You should get a `JsonParseAfterRedirectError` console error.

##### Test the change:
* Stash the change that was made to the `feature-utils.js` file in the previous steps
* Checkout the feature branch, apply the change from the stash and re-build the Jetpack plugin
* Now, making sure the protect plugin is not currently active, repeat the CLI testing steps from above. The return should be clean and the plugin should be activated
* Deactivate the protect plugin
* Now, repeat the steps above to activate the protect plugin from the recommendations summary screen. This time, you should not get any console errors or messages.
* Visit the plugins page, deactivate the protect plugin
* Now, from the plugins page, activate the protect plugin and confirm that you are redirected to the protect plugin page after activation - this is expected only when activating from the plugins screen.
* Test activating the protect plugin from the plugins screen as part of a bulk activation. You should NOT be redirected in this case.
